### PR TITLE
GH1199, Validate SharedSwapchains

### LINKED ISSUE #1199: Track swapchains created by CreateSharedSwapchainsKHR properly
Currently this function is intercepted, but the created objects are not tracked. Any further use of those swapchains will produce bad behavior in the layers.

### PR DESCRIPTION
**Pull request by mark-lunarg**
Validation was in place for parameter_validation and object_tracker, hooked up the existing CreateSwapchainKHR validation to CreateSharedSwapchainsKHR.  Fixes #1199.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -12309,9 +12309,43 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSharedSwapchainsKHR(VkDevice device, uint32
                                                          const VkSwapchainCreateInfoKHR *pCreateInfos,
                                                          const VkAllocationCallbacks *pAllocator, VkSwapchainKHR *pSwapchains) {
     layer_data *dev_data = get_my_data_ptr(get_dispatch_key(device), layer_data_map);
+    std::vector<SURFACE_STATE *> surface_state;
+    std::vector<SWAPCHAIN_NODE *> old_swapchain_state;
 
+    if (pCreateInfos) {
+        std::unique_lock<std::mutex> lock(global_lock);
```
Comment (jeremy-lunarg) [inline layers/core_validation.cpp]: Can use lock_guard here if so inclined.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -12309,9 +12309,43 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSharedSwapchainsKHR(VkDevice device, uint32
                                                          const VkSwapchainCreateInfoKHR *pCreateInfos,
                                                          const VkAllocationCallbacks *pAllocator, VkSwapchainKHR *pSwapchains) {
     layer_data *dev_data = get_my_data_ptr(get_dispatch_key(device), layer_data_map);
+    std::vector<SURFACE_STATE *> surface_state;
+    std::vector<SWAPCHAIN_NODE *> old_swapchain_state;
```
Comment (jeremy-lunarg) [inline layers/core_validation.cpp]: Do you want to separate into Pre/Post validate calls?

Comment (mark-lunarg) [inline layers/core_validation.cpp]: Ah, yeah, I forgot.  I'll add another commit to break up this and CreateSwapChains at the same time.  Thanks!

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -12309,9 +12309,43 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSharedSwapchainsKHR(VkDevice device, uint32
...
 
+    if (pCreateInfos) {
+        std::unique_lock<std::mutex> lock(global_lock);
+        for (uint32_t i = 0; i < swapchainCount; i++) {
+            surface_state.push_back(getSurfaceState(dev_data->instance_data, pCreateInfos[i].surface));
+            old_swapchain_state.push_back(getSwapchainNode(dev_data, pCreateInfos[i].oldSwapchain));
+            if (PreCallValidateCreateSwapchainKHR(dev_data, &pCreateInfos[i], surface_state[i], old_swapchain_state[i])) {
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: It would be nice for the errors to indicate which index was bogus.

**Comment by mark-lunarg** [comment]
Updated for all review feedback, pushed to Master.

**Comment by chrisforbes** [comment]
@mark-lunarg this makes CreateSharedSwapChains behave well, but have you looked at whether CV does something sensible with the shared images? I suspect you'll find some breakage there.

**Comment by mark-lunarg** [comment]
Yeah, I skimmed over that paragraph.  Sigh.  This should at least cover the existing implemented validation for swapchains, but there is some additional validation specifically for the shared ones.

**Pull request closed without merging**